File: Structural_Patterns/Composite/cpp/Composite.cpp
/*!
 *  \file       Composite.cpp
 *  \brief      
 *  
 */


#include "Composite.hpp"


/*!
 *  \note   A composite object can add or remove other components (both simple or complex) to or from its child list.
 */
void Composite::add(Component *component)
{
    m_children.push_back(component);
    component->setParent(this);
}

void Composite::remove(Component *component)
{
    m_children.remove(component);
    component->setParent(nullptr);
}

bool Composite::isComposite() const
{
    return true;
}

/*!
 *  \attention  The Composite executes its primary logic in a particular way.
 *              It traverses recursively through all its children, collecting and summarizing their results.
 *              Since the composite's children pass these calls to their children, and so forth,
 *              the whole object tree is traversed as a result.
 */
std::string Composite::operation() const
{
    std::string result;

    for (const Component *comp : m_children)
    {
        /*!
         *  \note   This loop will traverse to access each comp,
         *          if comp is Composite* object, it will call Composite::operation() recursively,
         *          because Composite* object has the m_children of list<Component *> container
         * 
         *          if comp is Leaf* object, it will then call Leaf::operation(), but not recursively,
         *          because Leaf* object has no any children, it is at the bottom.
         */
        if (comp == m_children.back())
        {
            result += comp->operation();
        }
        else
        {
            result += comp->operation() + "+";
        }
    }
    return "Branch(" + result + ")";
}

File: Structural_Patterns/Composite/c#/Composite.cs
/*!
 *  \file       Composite.cs
 *  \brief      
 *  
 */


using System;
using System.Collections.Generic
using DesignPatterns.StructuralPatterns.Composite.Component


namespace DesignPatterns.StructuralPatterns.Composite
{
    public class Composite : Component
    {
        protected List<Component> m_children = new List<Component>();

        public override void add(Component component)
        {
            m_children.add(component);
        }

        public override void remove(Component component)
        {
            m_children.remove(component);
        }

        public override string operation()
        {
            int i = 0;
            string result = "Branch(";

            foreach (Component component in m_children)
            {
                result += component.operation();
                if (i != m_children.Count() - 1)
                {
                    result += "+";
                }
                i++;
            }
            return result + ")";
        }
    }
}